kuka_kr3_support/urdf/kr3r540.xacro

```diff
@@ -1,13 +1,39 @@
 <?xml version="1.0" ?>
-<robot name="kuka_kr3r540" xmlns:xacro="http://wiki.ros.org/xacro">
+<robot name="kr3r540" xmlns:xacro="http://wiki.ros.org/xacro">
 
   <xacro:arg name="prefix" default="" />
+
   <xacro:arg name="use_mock_hardware" default="false" />
-  
+  <xacro:arg name="mock_sensor_commands" default="false" />
+  <xacro:arg name="sim_gazebo_classic" default="false" />
+  <xacro:arg name="sim_gazebo" default="false" />
+  <xacro:arg name="simulation_controllers" default="" />
+
+
   <xacro:include filename="$(find kuka_kr3_support)/urdf/kr3r540_macro.xacro"/>
-  <xacro:include filename="$(find kuka_resources)/urdf/common_macro.ros2_control.xacro"/>
+  <!-- <xacro:include filename="$(find kuka_resources)/urdf/common_macro.ros2_control.xacro"/> -->
+  <xacro:include filename="$(find kuka_resources)/urdf/common_sim_macro.ros2_control.xacro"/>
+
+  <!-- <xacro:kuka_common_ros2_control_macro name="kr3r540_mock" prefix="$(arg prefix)" use_mock_hardware="$(arg use_mock_hardware)"/> -->
+  <!-- <xacro:kr3r540_ros2_control name="kr3r540" prefix="$(arg prefix)" use_mock_hardware="$(arg use_mock_hardware)" simulation_controllers="$(arg simulation_controllers)"/> -->
+
+   <!-- create link fixed to the "world" -->
+  <link name="world" />
+
+  <!-- Load robot's macro with parameters -->
+  <!-- set prefix if multiple robots are used -->
+  <xacro:kr3r540 prefix="$(arg prefix)" parent="world" >
+    <origin xyz="0 0 0" rpy="0 0 0" />          <!-- position robot in the world -->
+  </xacro:kr3r540>
 
-  <xacro:kuka_common_ros2_control_macro name="kr3r540_mock" prefix="$(arg prefix)" use_mock_hardware="$(arg use_mock_hardware)"/>
+  <xacro:kr3r540_ros2_control
+    name="kr3r540"
+    prefix="$(arg prefix)"
+    use_mock_hardware="$(arg use_mock_hardware)"
+    mock_sensor_commands="$(arg mock_sensor_commands)"
+    sim_gazebo_classic="$(arg sim_gazebo_classic)"
+    sim_gazebo="$(arg sim_gazebo)"
+    simulation_controllers="$(arg simulation_controllers)" />
```

Comment:
these changes seem to hard-code `ros2_control` support and pre-configure the base models for use with Gazebo.

We've always tried to keep the base support packages as stand-alone and technology-agnostic as possible, and add simulator support in separate packages. After all: there are more simulators than Gazebo.

Was this a conscious design decision, or is this just 'a test'?

Reply:
I've thought about this a bit and it seems things are a bit more complex than you describe them here.

I'll just write down some bullets here, but if needed, I can go into more detail.

The following would be why I believe it would be better -- from a software engineering perspective -- to separate concerns a bit better:

 1. unless left out of the `package.xml`, adding `ros2_control` to the base models increases installation footprint. If I just want to visualise data, there would be no need to having to install `ros2_control` or its Gazebo part
 1. adding it here assumes all UDF parsers can deal with unknown elements. That's definitely not always the case and has been a pain to people using / writing such parsers in the past
 1. it mixes responsibilities: URDFs describe kinematics and some dynamics. Adding the `ros2_control` elements now adds information only useful for one possible control / interfacing option. I find it difficult to come up with a valid rationale for why that sort of information belongs in what is essentially a robot modelling language
 1. if a use-case doesn't need `ros2_control`, the URDF or xacro is now (potentially) much larger and contains unneeded stuff which complicates the model
   1. one good example of this would be if there are several available drivers, which don't use `ros2_control`. This has happened in the past, quite a few times actually. Why hard-code support for one particular dependency, which is not needed (as there are alternative ways of configuring things)?
 1. if parameters need to be changed, how would that be done without either changing what could essentially treated as read-only files, without either forking, branching or copying the original, then making the changes?

All of the above points also apply to hard-coding support for a particular simulator -- Gazebo in this case. There are many, many simulation environments capable of loading URDFs and XACROs. There are probably only 2 which support `ros2_control`. The rest either chokes on the custom elements, or doesn't use any of the information.

My suggestion would be to rely on *composability* more, which is exactly what XACRO was created for, and can address almost all of the points I list above.

Keep the base models as simple as possible: describe kinematics, dynamics and geometry.

Now add support for other functionality in separate packages. If you'd really want, you could wrap *all* base model XACROs in a single additional one which adds support for `ros2_control` and parameterise all of that by "model type" (or whatever name you want to use for the argument).

That would adhere to SE principles, improve reusability, help future users, improve maintainability *and* let you do what you're doing here.

Technically, you wouldn't even need to "port" any of the `.xacro`s, as what I suggest could already be done with the base `_macro.xacro`s now. Would potentially even be *less* work.

Reply:
@gavanderhoorn You are fully right! Thanks for opening my eyes!!!! Thanks for the great and detailed explanation!

@Robotgir please do the following:

1. Create a new package `kuka_ros2_control_support`.
2. Add there a general launch file that replaces all `test_ros2_control*.launch.py` files by defining robot type and model with arguments. Use `choice` on this argument to define a limited list of packages and arguments.
3. Add `.xacro` file there that combines robot's description and ros2_control part. The ros2_control part should be defined in this new package, not in `kuka_common`.
4. Add simulator support into a new package `kuka_ros2_control_sim_support` that adds simulator's parts on top of the ros2_control stuff. This is because the one that wants to use ros2_control driver doesn't have to install simulators.

**GOAL** - the packages should contain only description and have basic functionality for testing it.


@gavanderhoorn does this steps sound plausible to you? Have I missed something, or should we do something differently?

Reply:
This makes a lot of sense to me, yes @destogl.

> Have I missed something, or should we do something differently?

I believe the only additional suggestion I would have would be to be as clear as possible about what packages are adding / supporting, so following REP-144, I'd suggest renaming `kuka_ros2_control_sim_support` to something which mentions Gazebo and/or Ignition.

Perhaps `kuka_ros2_control_gazebo_support` would already be sufficient:

 - it's clear about the fact it uses `ros2_control`
 - it's clear about the fact it adds Gazebo support
 - and it's clear it's for "KUKAs" (whatever that means)

I expect people looking for Ignition support will still look at a package with just `gazebo` in the name, and the readme could clarify there's also Ignition support provided.

No point in confusing people by including the generic `sim` in the name of the package, as that doesn't tell you *which* simulator. Being clear about the simulator supported makes sense to me, as it would be obvious there'd be no point in trying to use the package with anything but Gazebo (so Webots users -- fi -- would have to look elsewhere).

One additional comment:

> Use `choice` on this argument to define a limited list of packages and arguments.

while this makes sense from one perspective (preventing users from trying to use the wrong arguments) it does incur maintenance overhead, as those `choice`s will now have to be updated whenever someone adds a new support package and/or variant to an existing one.

I'll leave it to you to decide whether it's worth that overhead (ie: one more round of reviews when someone contributes or updates a support package).